### DISCUSSION TITLE
how to implement auto layout with Dagre that will have subflows (or dynamic grouping)?

### DISCUSSION THREAD

**Discussion by mtruongsa** [xyflow/xyflow#2968] [Layouting] [answered]
I wonder if anyone has done this implementation before. I am stuck at how to put all child nodes inside a group node. Dagre library calculates the child nodes' position totally different than the group node. It makes my flow look like 2 separate flows and the group node is empty.

**Comment by chrtze** [comment, marked as answer]
hey @mtruongsa, it could be possible that you need to calculate the relative positions for the child nodes. Within React Flow, we are using relative coordinates to position the child nodes (e.g. `position: { x: 0, y: 0 }, parentNode: 'parent'` will be placed in the top left corner of `parent`).

When using a layout library, it might be that the coordinates are all relative to the viewport so you might to calculate the child positions like this:

```ts
x = child.x - parent.x;
y = child.y - parent.y
```

**Comment by elvin-hwang** [comment]
Hi, I'm having a similar issue but the above solution does not seem to have solved it for me. 
@mtruongsa if you are able, are you able to share the solution that worked for you in this case?

This is my `getLayoutedElements` function that I slightly modified from this example https://reactflow.dev/docs/examples/layout/dagre/. The only difference from the example might be that I set all the initial positions of my nodes to (0, 0).

```
const getLayoutedElements = (nodes: any, edges: any, direction = 'TB'): any => {
  const isHorizontal = direction === 'LR'
  dagreGraph.setGraph({ rankdir: direction })

  nodes.forEach((node: any) => {
      dagreGraph.setNode(node.id, {
        width: node.width,
        height: node.height,
  })

  edges.forEach((edge: any) => {
    dagreGraph.setEdge(edge.source, edge.target)
  })

  dagre.layout(dagreGraph)

  nodes.forEach((node: any) => {
    const nodeWithPosition = dagreGraph.node(node.id)
    node.targetPosition = isHorizontal ? 'left' : 'top'
    node.sourcePosition = isHorizontal ? 'right' : 'bottom'

    if (node.parentNode) {
      const parentNodeWithPosition = dagreGraph.node(node.parentNode)
      node.position = {
         x: nodeWithPosition.x - parentNodeWithPosition.x,
         y: nodeWithPosition.y - parentNodeWithPosition.y,
       }
    } else {
      node.position = {
        x: nodeWithPosition.x - node.width / 2,
        y: nodeWithPosition.y - node.height / 2,
      }
    }

    return node
  })

  return { nodes, edges }
}
```

**Comment by atrbgithub** [comment]
@elvin-hwang Thanks for providing your code example, was you able to get it working?

After a bit of fiddling I ended up with this code:

```ts
    if (node.parentNode) {
      const parentNodeWithPosition = dagreGraph.node(node.parentNode)
      node.position = {
        x: nodeWithPosition.x - (parentNodeWithPosition.x - parentNodeWithPosition.width/2) - nodeWidth/2,
        y: nodeWithPosition.y - ( parentNodeWithPosition.y - parentNodeWithPosition.height/2) - nodeHeight/2,
      }
    } else {
      node.position = {
        x: (nodeWithPosition.x - nodeWithPosition.width/2),
        y: (nodeWithPosition.y - nodeWithPosition.height/2),
      }
    }
```

This seems to work ok for me.

**Reply by youngjunning** [reply]
Thanks for your great work.

**Comment by elvin-hwang** [comment]
I did solve that part of the issue, but now I'm facing an issue with calculating the width/height of my child flows. I will not know the full width/height of my child flow until after the initial render, but I'm not sure if there's an easy way to calculate this using react-flow other than just iterating through each of my sub flows. 

I may consider using another library for this use case

**Reply by iio888** [reply]
Hi, did you get great solution for this? @elvin-hwang
